File: fuzz/oracle.py
"""Python-side driver for the Node fuse.js oracle.

Spawns ``fuzz/oracle.js`` once and talks newline-delimited JSON to it, so a
differential run costs one process for the whole session instead of one per
case.

This module is *test infrastructure only*. The shipped ``fusejs`` package has
no Node dependency and never imports anything from here.
"""

from __future__ import annotations

import json
import subprocess
from pathlib import Path
from types import TracebackType
from typing import Any

ORACLE_JS = Path(__file__).resolve().parent / "oracle.js"


class OracleError(RuntimeError):
    """The Node oracle rejected a command."""


class Oracle:
    """A long-lived fuse.js process answering queries over stdin/stdout."""

    def __init__(self, node: str = "node") -> None:
        self._proc = subprocess.Popen(
            [node, str(ORACLE_JS)],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            encoding="utf-8",
            bufsize=1,
            cwd=str(ORACLE_JS.parent),
        )

    def __enter__(self) -> Oracle:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        self.close()

    def close(self) -> None:
        """Shut the Node process down."""
        if self._proc.poll() is None:
            if self._proc.stdin is not None:
                self._proc.stdin.close()
            try:
                self._proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self._proc.kill()

    def call(self, **command: Any) -> Any:
        """Send one command and return its result."""
        stdin, stdout = self._proc.stdin, self._proc.stdout
        if stdin is None or stdout is None:
            raise OracleError("oracle process has no pipes")

        stdin.write(json.dumps(command, ensure_ascii=False) + "\n")
        stdin.flush()

        line = stdout.readline()
        if not line:
            stderr = self._proc.stderr.read() if self._proc.stderr else ""
            raise OracleError(f"oracle exited unexpectedly: {stderr.strip()}")

        response = json.loads(line)
        if not response.get("ok"):
            raise OracleError(response.get("error", "unknown oracle error"))
        return response["result"]

    # ── Convenience wrappers ───────────────────────────────────────

    def version(self) -> str:
        """The fuse.js version backing this oracle."""
        version: str = self.call(op="version")["version"]
        return version

    def match(self, pattern: str, text: str, options: dict[str, Any]) -> Any:
        """``Fuse.match(pattern, text, options)``."""
        return self.call(op="match", pattern=pattern, text=text, options=options)

    def search(
        self,
        docs: list[Any],
        query: Any,
        options: dict[str, Any],
        search_options: dict[str, Any] | None = None,
    ) -> Any:
        """``new Fuse(docs, options).search(query, searchOptions)``."""
        return self.call(
            op="search",
            docs=docs,
            query=query,
            options=options,
            searchOptions=search_options,
        )

    def create_index(
        self, keys: list[Any], docs: list[Any], options: dict[str, Any]
    ) -> Any:
        """``Fuse.createIndex(keys, docs, options).toJSON()``."""
        return self.call(op="createIndex", keys=keys, docs=docs, options=options)

File: fuzz/oracle.js
// Node-side oracle for the differential harness.
//
// Reads newline-delimited JSON commands on stdin and writes one
// newline-delimited JSON response per command on stdout. Keeping it a
// long-lived process (rather than one `node -e` per case) is what makes a
// 60-second fuzz run produce tens of thousands of cases instead of hundreds.
//
// This file is oracle-side only. Nothing here ships with the Python package —
// see DECISIONS.md on the "no source-language runtime" rule.

import { createInterface } from 'node:readline'
import Fuse from '../tests/original/dist/fuse.mjs'

const rl = createInterface({ input: process.stdin, terminal: false })

// Sort keys so the JSON text itself is comparable, not just the parsed value.
function canonical(value) {
  if (Array.isArray(value)) return value.map(canonical)
  if (value && typeof value === 'object') {
    return Object.fromEntries(
      Object.keys(value)
        .sort()
        .map((k) => [k, canonical(value[k])])
    )
  }
  return value
}

function handle(cmd) {
  switch (cmd.op) {
    case 'version':
      return { version: Fuse.version }

    case 'match':
      return Fuse.match(cmd.pattern, cmd.text, cmd.options || {})

    // Batch `Math.pow`, for verifying the fdlibm port in src/pyfuse/_fdlibm.py.
    // V8's Math.pow is not correctly rounded, so the Python port has to
    // reproduce *its* rounding rather than the true result.
    //
    // Operands and results travel as little-endian hex float bits: exact in
    // both directions, and it sidesteps JSON's inability to carry Infinity
    // and NaN at all.
    case 'pow': {
      const buf = Buffer.allocUnsafe(8)
      const decode = (hex) => Buffer.from(hex, 'hex').readDoubleLE(0)
      return cmd.pairs.map(([b, e]) => {
        buf.writeDoubleLE(Math.pow(decode(b), decode(e)))
        return buf.toString('hex')
      })
    }

    case 'search': {
      const fuse = new Fuse(cmd.docs, cmd.options || {})
      return fuse.search(cmd.query, cmd.searchOptions || undefined)
    }

    case 'createIndex': {
      const index = Fuse.createIndex(cmd.keys, cmd.docs, cmd.options || {})
      return index.toJSON()
    }

    case 'parseQuery':
      // Strip the attached searcher instances; only the tree shape is
      // comparable across languages.
      return JSON.parse(
        JSON.stringify(Fuse.parseQuery(cmd.query, cmd.options || {}, { auto: false }))
      )

    default:
      throw new Error(`unknown op: ${cmd.op}`)
  }
}

rl.on('line', (line) => {
  if (!line.trim()) return
  let response
  try {
    response = { ok: true, result: canonical(handle(JSON.parse(line))) }
  } catch (err) {
    response = { ok: false, error: err.message }
  }
  process.stdout.write(JSON.stringify(response) + '\n')
})
